src/tmpfiles/tmpfiles.c

```diff
@@ -1261,7 +1261,9 @@ static int create_item(Item *i) {
                 log_debug("%s directory \"%s\".", creation_mode_verb_to_string(creation), i->path);
 
                 if (IN_SET(i->type, CREATE_SUBVOLUME_NEW_QUOTA, CREATE_SUBVOLUME_INHERIT_QUOTA)) {
-                        r = btrfs_subvol_auto_qgroup(i->path, 0, i->type == CREATE_SUBVOLUME_NEW_QUOTA);
+                        r = path_check_fstype(i->path, BTRFS_SUPER_MAGIC);
+                        if (r == 1)
+                                 r = btrfs_subvol_auto_qgroup(i->path, 0, i->type == CREATE_SUBVOLUME_NEW_QUOTA);
```

Comment:
> O_PATH on an autofs mount point will not trigger the mount.

Ah! good to know. Would be excellent if that could be documented in the man page, too...

> You could get btrfs_subvol_auto_qgroup() to open with O_PATH, check if it is btrfs, then if it is - openat(fd, ".", O_DIRECTORY|O_RDONLY...) to get an fd for ioctl.

Similar to that, couldn't we do an open("/proc/self/fd/%i") to convert an O_PATH fd to a non-O_PATH fd? or is that flag inherited in that case?

> The reason that I didn't offer code to check everything for autofs is that I think autofs should be as transparent as possible. In this case, it isn't that we don't want autofs, it is that we do want btrfs. 

Oh, not quite. Note that "q", "v" and "Q" degrade gracefully to "d" if btrfs isn't used, and the offending lines also carry 0755 as mode to apply to these dirs. Hence, regardless if btrfs is used or not, the mkdir and chmod will trigger any autofs set up this way... Hence yupp, I think the way to avoid autofs deadlocks is to check for autofs here...

So my current thinking would be to build on chase_symlinks() for this (chase_symlinks() is our reeimplementation of glibc's canonicalize_file_name() but can operate relative to some specified root directory). We could add a new flag to it, which would stop chasing as soon as a component that is an autofs mount point. In that case it would return a specific new error code, and tmpfiles.d would look for that before realizing a line and skip the line in that case.